src/main/java/org/datadog/jenkins/plugins/datadog/tracer/DatadogTracerConfigurator.java

```diff
@@ -0,0 +1,175 @@
+        }
+
+        DatadogGlobalConfiguration datadogConfig = DatadogUtilities.getDatadogGlobalDescriptor();
+        if (datadogConfig == null) {
+            LOGGER.log(Level.WARNING, "Cannot set up tracer: Datadog config not found");
+            return Collections.emptyMap();
+        }
+
+        TopLevelItem topLevelItem = getTopLevelItem(run);
+        FilePath workspacePath = node.getWorkspaceFor(topLevelItem);
+        if (workspacePath == null) {
+            throw new IllegalStateException("Cannot find workspace path for " + topLevelItem + " on " + node);
+        }
+
+        Map<String, String> variables = new HashMap<>(getCommonEnvVariables(datadogConfig, tracerConfig));
+        for (TracerLanguage language : languages) {
+            TracerConfigurator tracerConfigurator = configurators.get(language);
+            if (tracerConfigurator == null) {
+                LOGGER.log(Level.WARNING, "Cannot find tracer configurator for " + language);
+                continue;
+            }
+
+            try {
+                Map<String, String> languageVariables = tracerConfigurator.configure(tracerConfig, node, workspacePath, envs);
+                variables.putAll(languageVariables);
+            } catch (Exception e) {
+                LOGGER.log(Level.WARNING, "Error while configuring " + language + " Datadog Tracer for run " + run + " and node " + node, e);
+                return Collections.emptyMap();
+            }
+        }
+        run.addAction(new ConfigureTracerAction(node, languages, variables));
+        return variables;
+    }
+
+    private static TopLevelItem getTopLevelItem(Run<?, ?> run) {
+        if (run instanceof AbstractBuild) {
+            AbstractBuild<?, ?> build = (AbstractBuild<?, ?>) run;
+            AbstractProject<?, ?> project = build.getProject();
+            if (project instanceof TopLevelItem) {
+                return (TopLevelItem) project;
+            } else {
+                throw new IllegalArgumentException("Unexpected type of project: " + project);
+            }
+        } else {
+            Job<?, ?> parent = run.getParent();
+            if (parent instanceof TopLevelItem) {
+                return (TopLevelItem) parent;
+            } else {
+                throw new IllegalArgumentException("Unexpected type of run parent: " + parent);
+            }
+        }
+    }
+
+    private static Map<String, String> getCommonEnvVariables(DatadogGlobalConfiguration datadogConfig,
+                                                             DatadogTracerJobProperty<?> tracerConfig) {
+        Map<String, String> variables = new HashMap<>();
+        variables.put("DD_CIVISIBILITY_ENABLED", "true");
+        variables.put("DD_ENV", "ci");
+        variables.put("DD_SERVICE", tracerConfig.getServiceName());
+
+        DatadogClient.ClientType clientType = DatadogClient.ClientType.valueOf(datadogConfig.getReportWith());
+        switch (clientType) {
+            case HTTP:
```

Comment:
Is this with or without TLS?  If not, followup feels like it's out of scope but it does appear that there's some sensitive items being sent over the wire (API Key/App key).

Reply:
The injected APM tracer will always use HTTPS. The plugin itself, when sending job data to Datadog, will use whatever URL is supplied by the user in the config settings, so it depends on whether Datadog allows unencrypted HTTP connections or not. I tried configuring URLs with `http://` protocol, but the server side always seems to negotiate the usage of HTTPS.

Reply:
Regarding `will use whatever URL is supplied by the user in the config settings` - so we don't necessary enforce this has to go to a datadog domain, and correct me if I'm wrong: this tracer isn't designed to enforce that.  It provides the shape of data we expect, but we allow the user to send to any recipient?  Because of that, it's a design decision to use HTTP by default, where the recipient ends up negotiating encryption?

Reply:
The `HTTP` constant name might be misleading here, in reality the URL can be either `http` or `https`, depending on what the user configures.

Yes, technically the user can configure the plugin to send the data to any endpoint, not necessarily Datadog. I do not know if this is a valid or expected use case (might be useful if the user wants the data to go through some proxy), but it is technically possible to enter anything into the endpoint URL inputs: the validation logic simply checks that the url string is not empty and that it contains `http` substring.

The default values all point to Datadog endpoints, and use `https` protocol. If I change them and specify `http`, the remote end (Datadog backend) will still force the usage of `https`.

As a side note, URL config inputs, validation logic, and different client types were implemented before. In this feature I am using this pre-existing logic to ensure that the injected tracer uses the same config values as the plugin itself.